fix: set `difficulty` default to 1 and add `totalDifficulty`

eggplantzzz opened the pull request:
Currently Ganache sets all block headers to have `difficulty` and `totalDifficulty` hard-coded to 0. This PR causes Ganache to set the difficulty of each block (by default) to 1 (the user can specify their own block difficulty on the command line). The `totalDifficulty` is also now calculated based on the previous block's `totalDifficulty` and the block difficulty.

eggplantzzz committed:
Initialize with a block difficulty of 1 by default

eggplantzzz committed:
Hook up the rest of the plumbing for creating blocks with difficulty

eggplantzzz committed:
Use the options to determine difficulty and not the previous block header

eggplantzzz committed:
Actually calculate totalDifficulty using previous totalDifficulty and current block's difficulty

eggplantzzz committed:
Make difficulty option a number rather than a string

eggplantzzz committed:
Add a couple of simple tests for totalDifficulty and difficulty; update test

eggplantzzz committed:
Use constant for difficulty in test

davidmurdoch commented on src/chains/ethereum/options/src/miner-options.ts:
```diff
@@ -43,6 +43,16 @@ export type MinerConfig = {
 
+    /**
+     * Sets the block difficulty
+     * @default 1
```
```suggestion

     * @default 1
```

davidmurdoch commented on src/chains/ethereum/options/src/miner-options.ts:
```diff
@@ -173,6 +183,12 @@ export const MinerOptions: Definitions<MinerConfig> = {
+  difficulty: {
+    normalize: Quantity.from,
+    cliDescription: "Sets the block difficulty for the network.",
+    default: () => Quantity.from(1),
```
Can you use `RPCQUANTITY_ONE` instead of `Quantity.from(1)`? You can get it from `import { utils } from "@ganache/utils"` here?

davidmurdoch commented on src/chains/ethereum/options/src/miner-options.ts:
```diff
@@ -173,6 +183,12 @@ export const MinerOptions: Definitions<MinerConfig> = {
   },
+  difficulty: {
+    normalize: Quantity.from,
+    cliDescription: "Sets the block difficulty for the network.",
```
```suggestion
    cliDescription: "Sets the block difficulty.",
```

davidmurdoch commented on src/chains/ethereum/ethereum/src/blockchain.ts:
```diff
@@ -555,7 +557,9 @@ export default class Blockchain extends Emittery.Typed<
-      Quantity.from(timestamp)
+      Quantity.from(timestamp),
+      this.#options.miner.difficulty,
+      Quantity.from(0) // we start the totalDifficulty at 0
```
Use 
```suggestion
      RPCQUANTITY_ZERO // we start the totalDifficulty at 0
```

davidmurdoch committed:
fix block size calculation

I had to store totalDifficulty in a different way so it wasn't included
in the `raw` fields (which messes with the hash).

davidmurdoch commented on src/chains/ethereum/ethereum/tests/api/eth/getBlockByNumber.test.ts:
```diff
@@ -0,0 +1,35 @@
+import assert from "assert";
```
Can you add a tests that change the starting default difficulty.

Also, lets test the increasing `totalDifficulty` (both with and without the default difficulty).

eggplantzzz replied:
sounds good to me

davidmurdoch reviewed (changes requested):
Really Awesome work. I committed a change, if you wouldn't mind reviewing it.

The gist of the change I made was because:

 * `totalDifficulty` shouldn't be included in block size calculations
 * `totalDifficulty` shouldn't be included in the block hash (which it previously was, because it was included in the `raw` array)

It seems to work now, and as a bonus, I figured out why our block size calculations were always off a little (PR for that coming soon!)

I also requested some more tests :-D

Thanks

eggplantzzz committed:
Update src/chains/ethereum/options/src/miner-options.ts

Co-authored-by: David Murdoch <davidmurdoch@users.noreply.github.com>

eggplantzzz committed:
Update src/chains/ethereum/options/src/miner-options.ts

Co-authored-by: David Murdoch <davidmurdoch@users.noreply.github.com>

eggplantzzz committed:
Use constant and update comment

eggplantzzz committed:
Use constant for miner options default

eggplantzzz committed:
Update some tests

davidmurdoch commented on src/chains/ethereum/ethereum/tests/api/eth/getBlockByNumber.test.ts:
```diff
@@ -0,0 +1,99 @@
+
+    await provider.send("eth_sendTransaction", [{ from, to: from }]);
+
+    await provider.once("message");
```
If you don't need the transaction itself, you can do `await provider.send("evm_mine")` instead.

Oh, actually... now that #793 is merged (as of yesterday), you can do `await provider.send("evm_mine", [{blocks: number}])`, no (backwards :smile:) loop needed!

eggplantzzz replied:
Oh cool! I'll merge in `develop` and update this!

eggplantzzz replied:
I like deleting code.

davidmurdoch commented on src/chains/ethereum/ethereum/tests/api/eth/getBlockByNumber.test.ts:
```diff
@@ -0,0 +1,99 @@
+            const block = await provider.send("eth_getBlockByNumber", ["0x0"]);
+            assert.strictEqual(
+              block.totalDifficulty,
+              `0x${DEFAULT_DIFFICULTY}`
```
If `DEFAULT_DIFFICULTY` is ever greater than 9 in the future this test will fail.

Other ways of solving this:

```
assert.strictEqual(
  block.totalDifficulty,
  Quantity.from(DEFAULT_DIFFICULTY).toString()
);
```

or 

```
assert.strictEqual(
  parseInt(block.totalDifficulty, 16),
  DEFAULT_DIFFICULTY
);
```

or

```
assert.strictEqual(
  BigInt(block.totalDifficulty),
  DEFAULT_DIFFICULTY // change DEFAULT_DIFFICULTY to a BigInt
);
```

and lots of others, I'm sure.

davidmurdoch replied:
Can you also update other places that use the `0x{number}` pattern in these tests? Thanks!

eggplantzzz replied:
Sorry, I meant to update these but must have overlooked them. Updated!

eggplantzzz committed:
Use HEX

eggplantzzz committed:
Get rid of helper method and use evm_mine in tests

davidmurdoch commented on src/chains/ethereum/ethereum/tests/api/eth/getBlockByNumber.test.ts:
```diff
@@ -0,0 +1,93 @@
+            const numberOfBlocksToMine = Math.floor(Math.random() * 10 + 1);
+            await provider.send("evm_mine", [{ blocks: numberOfBlocksToMine }]);
+            const block = await provider.send("eth_getBlockByNumber", [
+              `0x${numberOfBlocksToMine}`
```
When `numberOfBlocksToMine` to mine is `10` it sends `"0x10"`, which is `16`

```suggestion
              `0x${numberOfBlocksToMine.toString(16)}`
```

davidmurdoch commented on src/chains/ethereum/ethereum/tests/api/eth/getBlockByNumber.test.ts:
```diff
@@ -0,0 +1,93 @@
+            const numberOfBlocksToMine = Math.floor(Math.random() * 10 + 1);
+            await provider.send("evm_mine", [{ blocks: numberOfBlocksToMine }]);
+            const block = await provider.send("eth_getBlockByNumber", [
+              `0x${numberOfBlocksToMine}`
```
```suggestion
              `0x${numberOfBlocksToMine.toString(16)}`
```

davidmurdoch reviewed (changes requested):
Two more little "suggestions" and it's good to go!

davidmurdoch committed:
Update src/chains/ethereum/ethereum/tests/api/eth/getBlockByNumber.test.ts

davidmurdoch committed:
Update src/chains/ethereum/ethereum/tests/api/eth/getBlockByNumber.test.ts
